charmstore.go

```diff
@@ -76,50 +76,77 @@ func (s *CharmStore) Get(curl *charm.URL) (charm.Charm, error) {
 	if curl.Series == "bundle" {
 		return nil, errgo.Newf("expected a charm URL, got bundle URL %q", curl)
 	}
+	path, err := s.archivePath(curl)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	return charm.ReadCharmArchive(path)
+}
+
+// GetBundle implements Interface.GetBundle.
+func (s *CharmStore) GetBundle(curl *charm.URL) (charm.Bundle, error) {
+	// The cache location must have been previously set.
+	if CacheDir == "" {
+		panic("charm cache directory path is empty")
+	}
+	if curl.Series != "bundle" {
+		return nil, errgo.Newf("expected a bundle URL, got charm URL %q", curl)
+	}
+	path, err := s.archivePath(curl)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	return charm.ReadBundleArchive(path)
+}
 
-	// Prepare the cache directory and retrieve the charm.
+func (s *CharmStore) archivePath(curl *charm.URL) (string, error) {
```

Comment:
Perhaps a doc comment here would be appropriate:

// archivePath returns a local path to the downloaded archive
// of the given charm or bundle URL, storing it in CacheDir,
// which it creates if necessary. If an archive with a matching
// SHA hash already exists locally, it will use the local version.

?